Bluesky Social (Independent Publisher)

krautrocker opened the pull request:
---
### When submitting a connector, please make sure that you follow the requirements below, otherwise your PR might be rejected. We want to make you have a well-built connector, a smooth certification experience, and your users are happy :) 

If this is your first time submitting to GitHub and you need some help, please sign up for this [session](https://forms.office.com/pages/responsepage.aspx?id=KtIy2vgLW0SOgZbwvQuRaXDXyCl9DkBHq4A2OG7uLpdUMTFJWFFGVUxBNUFZQjZWRUdaOE5BMFkwNS4u). 

- [x] I attest that the connector doesn't exist on the Power Platform today. I've verified by checking the pull requests in GitHub and by searching for the connector on the platform or in the documentation. 
- [x] I attest that the connector works and I verified by deploying and testing all the operations. 
- [x] I attest that I have added detailed descriptions for all operations and parameters in the swagger file.
- [x] I attest that I have added response schemas to my actions, unless the response schema is dynamic. 
- [x] I validated the swagger file, `apiDefinition.swagger.json`, by running `paconn validate` command.
- [x] If this is a certified connector, I confirm that `apiProperties.json` has a valid brand color and doesn't use an invalid brand color, `#007ee5` or `#ffffff`. If this is an independent publisher connector, I confirm that I am not submitting a connector icon.

If you are an Independent Publisher, you must also attest to the following to ensure a smooth publishing process:
- [x] I have named this PR after the pattern of "Connector Name (Independent Publisher)" ex: HubSpot Marketing (Independent Publisher)
- [x] Within this PR markdown file, I have pasted screenshots that show: 3 unique operations (actions/triggers) working within a Flow. This can be in one flow or part of multiple flows. For each one of those flows, I have pasted in screenshots of the Flow succeeding. 
- [x] Within this PR markdown file, I have pasted in a screenshot from the Test operations section within the Custom Connector UI.
- [x] If the connector uses OAuth, I have provided detailed steps on how to create an app in the readme.md.

krautrocker committed:
Add Bluesky API (Independent Publisher) connector files

krautrocker commented:
@microsoft-github-policy-service agree [company="{Dan Romano aka IDR, Inc. dba SWOLCAT}"]

krautrocker commented:
@microsoft-github-policy-service agree company="Dan Romano aka IDR, Inc. dba SWOLCAT"

krautrocker committed:
Updated swagger file to include default responses and added .gitignore

troystaylor commented:
Hi Dan, I have a connector I’m working on too. Are you building more out or is this what you are going to submit (without the Proposal)?

krautrocker committed:
- Updated connector title and description to jive with standards

krautrocker commented:
> Hi Dan, I have a connector I’m working on too. Are you building more out or is this what you are going to submit (without the Proposal)?

Hi Troy, I will be building out more. These three actions are just the start and will be expanding in the next few hours. To get this up and running, we only need three actions but there's a lot more than we can and will add, naturally. What do you think?

troystaylor commented:
I tend to build them all. Are you having any problems with the OAuth?

krautrocker commented:
Right now, I've got basic auth going with the app passwords and rooting around for OAuth...but not seeing where we can create a client and secret. Posted in the dev channel on Bluesky but no response yet (recent post).

krautrocker committed:
- Updated security definitions to Basic, not API Key

krautrocker committed:
- Added new paths

krautrocker committed:
- Added new path to get hydrated feed

krautrocker committed:
- Added more paths

fordosa90 commented:
@krautrocker Dan, please remove changes to the BkkFutar connector, I am not yet planning to update its authentication ;)

krautrocker commented:
> @krautrocker Dan, please remove changes to the BkkFutar connector, I am not yet planning to update its authentication ;) 

Damn, my mistake - will remove at once this evening.

krautrocker committed:
Revert "- Updated security definitions to Basic, not API Key"

This reverts commit ea3bcd3c64d92a194ffef31cf1626b4a5329dd68.

krautrocker committed:
- Added more endpoints

krautrocker committed:
- Added additional endpoints

krautrocker committed:
- Updated readme.md

krautrocker committed:
- Added admin endpoints

- Updated readme file

krautrocker committed:
- Doc format

- Removed endpoints temporarily, adding back tomorrow

krautrocker committed:
- Changed basic auth to bearer token auth

krautrocker committed:
- Overhaul after convo with Troy Taylor

krautrocker committed:
- Change contact email, placed in readme
- Modified endpoints in readme

krautrocker committed:
- Adjusted endpoints as per paconn validator

krautrocker committed:
- Further updates to conform

krautrocker committed:
- Updated title to remove a forbidden word

krautrocker committed:
- Resolved all errors via paconn validate in swagger file

krautrocker commented:
Test operation validation below. Proof of life:

**1.) Create session**

![Bluesky_Custom_Connector_CreateSession](https://github.com/user-attachments/assets/d1257beb-82e1-4909-9b31-2c21ac64255b)

**2.) Get preferences**

![Bluesky_Custom_Connector_GetPreferences](https://github.com/user-attachments/assets/2fa224b8-b8fa-46a4-83fc-35ad2f5ef616)

**3.) List notifications**

![Bluesky_Custom_Connector_ListNotifications](https://github.com/user-attachments/assets/598ee41b-57ad-4696-9c9c-d2bb2fb570bc)

/**/

Flow will be ready later tonight or tomorrow once warnings and schemas are otherwise validate. All errors have been cleared via paconn validate.

krautrocker commented:
Working flow:

Actions list (non-exhaustive):

![image](https://github.com/user-attachments/assets/2673d255-ff27-4494-a607-b8122d279759)

1.) Create session

![image](https://github.com/user-attachments/assets/49103762-135b-4d98-b059-62bd283e0710)

/**/

More tomorrow!

Laskewitz commented:
Well done, @krautrocker! 👌👏

krautrocker commented:
> Well done, @krautrocker! 👌👏

Thanks for the support. I've got some work left to do. High ground soon.

krautrocker commented:
Three actions working - Create session, get a user's profile, and retrieve a user's followers.

![image](https://github.com/user-attachments/assets/6caa290a-5609-4adf-9fba-5f0c4ff6f23e)

krautrocker committed:
- Swagger refinement

krautrocker committed:
- Additional endpoints

krautrocker committed:
- Removed file

krautrocker committed:
- Initial commit for Bluesky Control Panel (temporary name) to include actions for ATProto, chat, ozone

krautrocker committed:
- WOOF

krautrocker committed:
- Endpoints added

vmanoharas commented:
Hello @krautrocker, 

There is another PR opened with the same title (https://github.com/microsoft/PowerPlatformConnectors/pull/3711/files#diff-b0834cb083f9dfdbce3924041d0f322c218992fab6d7c968d8ec50b190ae5a25)
Are they same? if so, please submit with different title. Additionally, follow this link and submit your IP connector. The files are under gitignore. please check. 

https://github.com/microsoft/PowerPlatformConnectors

![image](https://github.com/user-attachments/assets/8df9fc6b-6f05-4ff5-92a5-cc56b029734a)

krautrocker commented:
> Hello @krautrocker, 
> 
> 
> 
> There is another PR opened with the same title (https://github.com/microsoft/PowerPlatformConnectors/pull/3711/files#diff-b0834cb083f9dfdbce3924041d0f322c218992fab6d7c968d8ec50b190ae5a25)
> 
> Are they same? if so, please submit with different title. Additionally, follow this link and submit your IP connector. The files are under gitignore. please check. 
> 
> 
> 
> https://github.com/microsoft/PowerPlatformConnectors
> 
> 
> 
> ![image](https://github.com/user-attachments/assets/8df9fc6b-6f05-4ff5-92a5-cc56b029734a)
> 
> 

Hiya @vmanoharas They are not the same and they should have two different names now.

Bluesky - app.bsky endpoints
Bluesky Control Panel - All other endpoints

Unfortunately, they can't all fit into one due to current limitations.

I will fix up the .gitignore and file placement, yikes. Thanks!

krautrocker committed:
Remove unnecessary .gitignore files

krautrocker committed:
- Updated endpoints

- .gitignore removal

krautrocker committed:
Delete .gitignore

Delete .gitignore

krautrocker commented:
@vmanoharas Alrighty, I removed the .gitignore file and I'm in the process of consolidating the folders for ease of review and clarity.

Thanks for the help, will have this finished ASAP for review next week.

krautrocker committed:
- Additional endpoints for app.bsky, pre-paconn validation

krautrocker committed:
- Added additional endpoints

- Added request responses

- Updated definitions for references within endpoints

- Validated with paconn, no errors

krautrocker committed:
Merge branch 'bluesky-social-connector' of https://github.com/krautrocker/PowerPlatformConnectors into bluesky-social-connector

krautrocker committed:
- Updated operation Ids to start with an upper-case letter

krautrocker committed:
- Updated folder name, IP not required in folder only PR title

krautrocker committed:
- Updated readme to include additional endpoints

krautrocker committed:
- Updated readme

- Validated with paconn, no errors

- Moved global definitions position

Ready to test with a client and secret.

krautrocker reviewed (commented):
Ready to test, but a few housekeeping items do remain. This PR is indeed different from this other [PR #3711](https://github.com/microsoft/PowerPlatformConnectors/pull/3711) but may merge the branches.

krautrocker committed:
- Authentication update, OAuth no dice

- Added connection parameters to apiProperties

- Added policy templates to apiProperties

- Added endpoint for refreshSession

- Validated with paconn, no errors

krautrocker reviewed (commented):
For this PR, I implemented connection parameters and policy templates per @Laskewitz guidance. I tested this in a flow and it works.

Additional testing required for authentication and, naturally, the other 70+ endpoints. actor, for example, is a parameter in about 12 endpoints that needed to be changed from actorId to just actor. QA + testing ready.

krautrocker committed:
- Fixed security definitions

krautrocker committed:
- Updated apiProperties

krautrocker committed:
- Further update to apiProperties

vmanoharas commented:
Hello @krautrocker, 

Can you please remove the word Proposal from the title?

krautrocker committed:
- Re-factored all endpoints. They should match the API docs + GitHub OpenAPI spec (though not 3.0)

- Removed dead code for definitions if they were only referenced but never uses.

- Combined all endpoints into a single connector. Not yet comprehensive but on the way to a one-stop shop.

- Validated with paconn, no errors.

krautrocker commented:
1.) Endpoints QA'd

Though not yet complete, I took a different approach today to effectively align the endpoints with Bluesky's specs. Parameters are matching up in both name and composition. 

2.) All-in-one

I am combining all namespaces and endpoints into a single connector. So far, so good. If this goes well, then there will be only one PR.

3.) Connector size

Small, only 150KB so far. The definition references really do save space, and there's probably more room for enhancement here. Global definitions have been a great start, which has cut down on code bulk.

4.) Removed dead code

A few definitions were "declared but never used." These have been removed.

/**/

Plan for tomorrow:

- Add remaining endpoints.

Plan after tomorrow:

- Test with data from Bluesky. Aside from the simple things such as actor Id, user email, I'll find values for a repo Id or any other endpoint in order to test all actions.

- Request review.

krautrocker committed:
- Added additional endpoints.

- Added 401 to responses.

- Validated with paconn, no errors.

krautrocker commented:
Only 50 endpoints remaining. Validation is clean. The rest will be completed tomorrow.

File size is far below the 1MB limit.

In order to test this, I will need play around in Bluesky. Create lists, get a post Id or other item and conduct operations on it via endpoints.

krautrocker committed:
- Added additional endpoints.

- Removed dead code, ProfileViewBasic.

- Added definition for 401 errors.

- Validated with paconn, no errors.

krautrocker committed:
- Futzed around with the multipart/form-data endpoints - uploadVideo, getBlob, and uploadBlob.

- Validated with paconn, no errors.

- Custom connector editor UI returns no errors or warnings.

krautrocker commented:
Notes:

- All endpoints are present in the connector. Around ~165 with a filze size < 500kb.

- Response codes need to be made consistent. This is mostly the error codes 400, 401, 403 and 500.

- Descriptions for parameters and endpoints can be added where they are missing and preferred.

Housekeeping items to be completed tomorrow.

/**/

This is ready to test. The app.bsky endpoints will be straight-forward. The "control panel" endpoints may require some thought but none of them seem supernatural.

krautrocker commented:
In the process of creating test cases for all endpoints. This will include happy paths and expected output for response codes.

🖖

vmanoharas commented on independent-publisher-connectors/Bluesky Social/apiDefinition.swagger.json:
```diff
@@ -0,0 +1,10863 @@
+          "app.bsky.actor"
+        ],
+        "summary": "Get Preferences for Current Account",
+        "description": "*This endpoint is part of the Bluesky application Lexicon APIs (app.bsky.*). Public endpoints which don't require authentication can be made directly against the public Bluesky AppView API: https://public.api.bsky.app.  Authenticated requests are usually made to the user's PDS, with automatic service proxying. Authenticated requests can be used for both public and non-publ ...
```
please keep all parameters descriptions short to fit in the text box. please see our [docs](https://learn.microsoft.com/en-us/connectors/custom-connectors/connector-string-guidance#parameters).

vmanoharas commented on independent-publisher-connectors/Bluesky Social/README.md:
```diff
@@ -0,0 +1,136 @@
+# Bluesky API Connector
```
Please update readme.md file with all the required fields mentioned in the below link. 

https://github.com/microsoft/PowerPlatformConnectors/blob/dev/templates/Independent%20Publisher/readme.md

vmanoharas commented on independent-publisher-connectors/Bluesky Social/apiProperties.json:
```diff
@@ -0,0 +1,73 @@
+{
```
Please update apiProerities.json file as per the below template and schema. 

https://github.com/microsoft/PowerPlatformConnectors/blob/dev/templates/Independent%20Publisher/apiProperties.json

https://github.com/microsoft/PowerPlatformConnectors/blob/dev/schemas/paconn-apiProperties.schema.json

krautrocker replied:
@vmanoharas I've made updates here and will re-review shorlty to make sure they jive with the standards provided in the links. Thank you!

vmanoharas reviewed (commented):
Hello @krautrocker, 

Kindly address the review comments and please go through the below document and update all the artifacts related to independent publisher.

krautrocker committed:
- Updated readme.md to bring file in line with MS docs

krautrocker committed:
- Added supported operations to readme.md

krautrocker committed:
- Shortened descriptions as per MS guidelines. Batch 1 of 2.

krautrocker committed:
- Shortened endpoint descriptions as per MS guidelines. Batch 2 of 2.

- Shortened parameter descriptions for various endpoints throughout the swagger.

krautrocker committed:
- Added helpful definitions to readme

krautrocker committed:
- Updated apiProperties to improve auth flow and bring in line with MS guidelines.

krautrocker committed:
- Updated parameter names, ongoing

krautrocker commented:
Recent work:

- apiProperties brought in line (pending review)
- Readme updated to include all endpoints and match guidelines.
- Authentication modified to inject bearer token into each request - this was more manual before

Ongoing work:

- Updating parameter names and response schemas to match docs
- Adding limit and cursor parameters (where supported)
- Default values

Moving forward, refining as we iterate.

krautrocker committed:
- Updated endpoint parameters and response schemas, ongoing

krautrocker committed:
- Updated endpoint definitions and response schemas

- Validated with paconn, no errors

krautrocker committed:
- Updated endpoint parameters and response schemas

- Validated with paconn, no errors

krautrocker committed:
- Updated endpoints and response schemas

- Validated with paconn, no errors

krautrocker committed:
- Updated endpoints and response schemas for remaining bsky.app

- Validated with paconn, no errors

krautrocker commented:
All set, continuing work on the endpoints.

krautrocker committed:
- Updated endpoints and response schemas for chat.bsky, ongoing

- Validated with paconn, no errors

krautrocker committed:
- Updated endpoints and response schemas

- Validated with paconn, no errors

vmanoharas commented:
Hi @krautrocker, 

Are you still working on this PR? if not please let us know to proceed further for approval process.

krautrocker commented:
Ah, I am indeed @vmanoharas. In the process of modifying endpoints. Should be another day or two. I'm hoping to be done with all updates by Friday.

krautrocker commented:
FYI - 512 object limit reached for app.bsky endpoints but I may have figured something out for the response. paconn validated without error so further testing will tell us more.

troystaylor commented:
@krautrocker As I have mentioned before, if you do not split the endpoints, you can use dynamic schema for the responses (as long as each response is no longer than 512 objects).

krautrocker commented:
@troystaylor Splitting the endpoints into separate connectors was discussed. Are you referring to x-ms-dynamic-schema?

troystaylor commented:
> @troystaylor Splitting the endpoints into separate connectors was discussed. Are you referring to x-ms-dynamic-schema?

Yes, you would need to use that with custom code for the responses built as C# dictionaries. Otherwise, it will have to be two connectors (until there is OpenAPI 3.0 support).

krautrocker commented:
> > @troystaylor Splitting the endpoints into separate connectors was discussed. Are you referring to x-ms-dynamic-schema?
> 
> Yes, you would need to use that with custom code for the responses built as C# dictionaries. Otherwise, it will have to be two connectors (until there is OpenAPI 3.0 support).

paconn is still validating without the 512 error...but maybe we need to agree on an approach.

krautrocker commented:
> Hi @krautrocker, 
> 
> 
> 
> Are you still working on this PR? if not please let us know to proceed further for approval process. 

@vmanoharas Still working on this, we're considering some options. I'm going to be out of pocket for a few days on vacation so will check back in once I return next week.

krautrocker committed:
- Reworked endpoints to align with spec, set 1

- paconn validated, no errors or 512 max limi showstoppers

krautrocker committed:
- Reworked endpoints to align with spec

- paconn validated, no errors or limits breached.

krautrocker committed:
- Reworked endpoints to align with spec

- paconn validated, no errors or limits breached

krautrocker committed:
- Reworked endpoints to align with spec

- paconn validated, no errors or limits breached.

krautrocker committed:
- Reworked endpoints to align with spec

- paconn validated, no errors or limits breached.

krautrocker committed:
- Reworked endpoints to align with spec.

- paconn validated, no errors or limits breached.

krautrocker committed:
- Reworked endpoints to align with spec

- paconn validated, no errors or limit breaches.

krautrocker committed:
- Reworked endpoints to align with spec.

- paconn validated, no errors or limit breaches.

krautrocker committed:
- Reworked endpoints to align with spec.

- paconn validated, no errors or limits breached.

krautrocker committed:
- Reworked endpoints to align with spec

- paconn validated, no errors or limits breached.

krautrocker committed:
- Reworked endpoints to align with spec

- paconn validated, no errors or limits breached.

krautrocker committed:
- Reworked endpoints to alight with spec.

- paconn validated, no errors or limits breached.

krautrocker committed:
- Reworked endpoints to align with spec.

- paconn validated, no errors or breached limits.

krautrocker committed:
- Reworked endpoints to align more with spec

- paconn validated, no errors or limits breached.

krautrocker committed:
- Reworked endpoints to align with spec

- paconn validated, no errors or limits breached.

krautrocker commented:
We have been to the edge and back. All endpoints are now included in the definition file. There's only a small amount of work left to do now.

Some thoughts on the work done over the past few weeks:

- **Re-understanding.** This API is more involved than I thought. It looks very different from a few weeks ago, as I realized my original approach required fundamental readjustment. I’m more confident in both the representation of the source API definition as well as my understanding of how this beast actually works. The gang’s all here, or at least far more than before.

- **512 object limit.** The language in the [MS Learn docs around schema limits](https://learn.microsoft.com/en-us/connectors/custom-connectors/faq) is subtle.

![image](https://github.com/user-attachments/assets/62d00661-30db-4e74-b093-d27f6d6fbbe8)

512 *per body.* If the object limit has been reached, paconn will throw an error, and it's a showstopper. @troystaylor provided guidance to use x-ms-dynamic-schema, and this seems to be the workaround here. Flattening the schema may work as well but x-ms-dynamic-schema is a strategically better. This only seems to pop up with the app.bsky namespace.

- **Conversion from OpenAPI to Swagger.** This part slowed things down considerably, as OpenAPI 3 isn't supported yet. Working with a generator presented issues on its own, namely around completeness and, sometimes, consistency. x-ms-summary, description, for example, were left out so a manual review has required greater care in the finer parts. Models sometimes get lazy, so generation may be fast but you must play the role of reviewer to ensure nothing was left out or cross-pollination occurs between Swagger and OpenAPI 3 markup.

- **Paconn CLI validation and custom connector UI.** The paconn CLI has a different tolerance level than the custom connector UI. Warnings would be present in one but not the other (capitalizing operation Id, for example). That added some overhead to the process, reconciling the two to create a synthesized product.

At this point, I'm reviewing some finer points.

- [x] Consistent error responses. Unauthorized vs Unauthorized - [Message]
- [x] Definition references. getRecord, for example, exists as a definition in multiple namespaces (sync, repo, moderation) so need to ensure they are correctly referenced in endpoint operations if they are not fully qualified. Only six instances, not a big to-do.
- [x] listConvos endpoint throws a weird warning (not an error) in paconn.
- [x] Descriptions are short and sweet.

@vmanoharas This is ready for further review.

krautrocker committed:
- Shortened some descriptions without sacrificing too much detail.

- paconn validated, no errors.

krautrocker committed:
- Added security definition to endpoints (some were missing)

- Standardized error responses

- Updated 200 message responses

- paconn validated, no errors

krautrocker commented:
Updating definition references tomorrow:

- [x] com.atproto.server.createAppPassword > createAppPassword
- [x] com.atproto.server.listAppPasswords > listAppPassword
- [x] com.atproto.label.subscribeLabels.info > subscribeLabelsInfo
- [x] com.atproto.sync.subscribeRepos.info > subscribeReposInfo
- [x] app.bsky.feed.describeFeedGenerator.links > describeFeedGeneratorLinks
- [x] com.atproto.server.describeServer.links > describeServerLinks
- [x] app.bsky.actor.defs.profileViewBasic > appProfileViewBasic
- [x] chat.bsky.actor.defs.profileViewBasic > chatProfileViewBasic
- [x] app.bsky.feed.defs.replyRef > defReplyRef
- [x] app.bsky.feed.post.replyRef > postReplyRef
- [x] app.bsky.embed.external.view > externalView
- [x] app.bsky.embed.images.view > imagesView
- [x] app.bsky.embed.record.view > recordView
- [x] app.bsky.embed.recordWithMedia.view > recordWithMediaView
- [x] app.bsky.embed.video.view > videoView
- [x] app.bsky.actor.defs.viewerState > actorViewerState
- [x] app.bsky.feed.defs.viewerState > feedViewerState

TTD:

1. Add definitions to Swagger
2. Update definition references in Swagger
3. Run paconn to validate
4. Commit

krautrocker committed:
- Updated definition references - serverAppPassword/listAppPassword and appProfileViewBasic/chatProfileViewBasic

- paconn validated, no errors

krautrocker committed:
- Updated definition references - feedReplyRef/postReplyRef

- paconn validated, no errors

krautrocker committed:
- Updated definitions and references, remaining batch.

- paconn validated, no errors.

krautrocker committed:
- Fixed listConvos x-dynamic-schema validation error

krautrocker commented:
Custom connector UI returns a "cannot validate schema" warning for about 19 endpoints and 14 definitions. Digging into this, this can be attributed to nested references ($ref). I can add the schema definition inline so the connector UI can parse and validate. This affects a small number of endpoints.

It's a handful of endpoints referencing definitions so if we put them inline, this may settle this...though I'm not crazy about adding definitions inline. $ref is cleaner and more organized but I would like validation to be as clean as possible.

The connector does indeed save in the UI, so this isn't a showstopper (at least not at the moment).

krautrocker committed:
- Made first letter of operation Ids upper-case to pass validation warnings in UI

krautrocker commented:
Work remaining:

- [x] Update policy parameters in apiProperties to acquire and use an access token (been chasing this one)
- [ ] Replace nested schemas in endpoints where schema cannot be validated.

krautrocker committed:
- Updated apiProperties to get and use token

- Modified the ImportRepo endpoints based on paconn validation

- paconn validated, no errors.

krautrocker commented:
Setting up a PDS this weekend. Stay tuned @Laskewitz

krautrocker commented:
PDS setup complete. It is online and accessible via web front end. DNS is resolving, allowing at least a half hour.

Next:

1.) We will connect to the instance via Power Automate

Then:

2.) Create a flow to test session token creation and refresh, confirming joy on the authentication mechanism.

From there:

3.) We will be in a position to test the rest of the endpoints with the connector.

The testing schedule shall commence per namespace:

- App - Typical social features such as liking a post, fetching a profile, creating a starter Pack, etc.
- Chat - DMs
- ATProto - Actions around server and account management
- Ozone - Moderation service

Human users have been engaged to help create lists, starter packs, send DMs, upload videos and memes, etc.

Success criteria:

- All endpoints operations are successful.

- Any problematic endpoints will be quarantined and set aside but going for 100% inclusion.

- Evidence of success will be provided via image.

Target UAT completion date: January 19th

krautrocker committed:
- Added runPolicyOn property to apiProperties.

- Validated with paconn, no errors.

- Changed security definition to "api_key" from "bearerAuth"

krautrocker commented:
Made quite a dent but a few things have tripped up the effort.

1.) TikTok still in play.

We've been monitoring the situation around social media platforms closely over the past few weeks. TikTok has gone dark officially but we expect it to return based on comments from the incoming administration. Palace intrigue indeed, and the situation remains fluid.

2.) Testing is 25% complete.

I would like to post figures to show testing progress, which is currently at 25%. Figures below the following comments:

- Despite setting up a PDS (Personal Data Server), DNS has still not resolved since it was set up last week. This means handles will display "@invalid-handle" until resolution. Good news is the email service is working and email and handle creations and updates are processed successfully via tokens in messages. A bit of a head scratcher after we've placed a TXT record at the registrar level...

A few discussions in forums says this goes away on its own but something else is going on here.

- It may be tricky to independently test ATProto and Ozone endpoints until the PDS issues are resolved. The next option is to use Postman to simulate PDS behavior but doesn't beat the real thing, naturally.

- Dynamic hosts are required. This became evident when we started testing the chat endpoints; errors were returned saying "XRPC not supported," which was anticipated. See here: https://docs.bsky.app/docs/advanced-guides/api-directory

apiProperties will be updated to handle dynamic hosts:

Host: {{host}}

Hosts will change dynamically at runtime; a particular flow construction must be used. When calling endpoints that don't use the bsky.social default endpoint, a flow needs to change the host to, say, the chat service, api.bsky.chat.

/**/

- app.bsky endpoints are close to 90% done.

- Chat and other services will be completed once dynamic hosts are implemented.

- ATProto and Ozone will be completed once the PDS DNS issue is resolved or a mock server is used in Postman.

More tomorrow and the rest of this week.

vmanoharas commented:
Hello @krautrocker, 

Is it ready for review? if not, we would request you to close this PR and submit when the connector is ready

krautrocker commented:
@vmanoharas I'm actively working on it. Parts of it are ready and I'm updating the PR as I go along. What's the policy here?

troystaylor commented:
@vmanoharas Wouldn't it be better for Dan just to restore the [Proposal] tag to the PR title? There's no reason to disconnect his existing commits

krautrocker commented:
Again, what's the policy?

krautrocker committed:
- Added sethost parameter to apiProperties to allow dynamically switching host for different endpoints

krautrocker committed:
- Fixed operation Id case for some operations.

krautrocker commented:
@vmanoharas All set for now. Please proceed.

krautrocker committed:
- Added expression for dynamic host

- Validated with paconn, no errors.

vmanoharas reviewed (dismissed):
Dear Partner

I hope you are doing well.
Congratulations, your pull request is approved. We will move forward with the certification process. We are preparing your connector for production deployment and will queue up for the next deployment schedule. Once your connector onboards the next deployment schedule, it starts to deploy your connector in our production environments that typically takes 4-5 weeks.
Please let us know if you have any questions.
Thank you very much for working with us.

vmanoharas commented:
[[certify-connector]]

Laskewitz commented:
[heart]         Daniel Laskewitz reacted to your message:
________________________________
From: vmanoharas ***@***.***>
Sent: Wednesday, January 29, 2025 3:54:28 PM
To: microsoft/PowerPlatformConnectors ***@***.***>
Cc: Daniel Laskewitz ***@***.***>; Mention ***@***.***>
Subject: Re: [microsoft/PowerPlatformConnectors] Bluesky Social (Independent Publisher) (PR #3706)


@vmanoharas approved this pull request.

Dear Partner

I hope you are doing well.
Congratulations, your pull request is approved. We will move forward with the certification process. We are preparing your connector for production deployment and will queue up for the next deployment schedule. Once your connector onboards the next deployment schedule, it starts to deploy your connector in our production environments that typically takes 4-5 weeks.
Please let us know if you have any questions.
Thank you very much for working with us.

—
Reply to this email directly, view it on GitHub<https://github.com/microsoft/PowerPlatformConnectors/pull/3706#pullrequestreview-2581494981>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AA4XUO6DBHI3JGU4WOOZYOT2ND2TJAVCNFSM6AAAAABRQQQTVKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKOBRGQ4TIOJYGE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

troystaylor commented:
Hi Dan,
Just now getting a chance to test your apiProperties and I'm seeing a couple issues with your policies:

1. Set Host URL - must use templateId called dynamichosturl. The parameter must be changed to x-ms-apimTemplateParameter.urlTemplate.  The urlTemplate value you've provided is an if-else statement which isn't valid.
2. Acquire Access Token - I can find no reference that this is a valid template, either in the [paconn Schema](https://github.com/microsoft/PowerPlatformConnectors/blob/b69cdf7e7520189ad5dd140df9a3dcc7ff44cea6/schemas/paconn-apiProperties.schema.json) or by online search. The tokenBody uses a key called username whereas the API requires the key to be [identifier](https://docs.bsky.app/docs/api/com-atproto-server-create-session). The key runPolicyOn is also invalid and must be included in the parameters as x-ms-apimTemplate-policySection set to Request.
3. The {{accessJwt}} would have to be set as a connection parameter. And runPolicyOn is also invalid as above.

Please correct me if I'm wrong with reference sources. I've also reached out to Lane Swenka to see if these policies are in development. @vmanoharas did the apiProperties fail your validator?

krautrocker committed:
- Set host URL - Changed templateId to dynamichosturl (per required standard). Updated x-ms-apimTemplateParameter.host to x-ms-apimTemplateParameter.urlTemplate.

- Removed acquiretoken (not a valid template).

- identifier now used instead of username in authentication.

- runPolicyOn removed; replaced with x-ms-apimTemplate-policySection: Request.

- accessJwt moved to connection parameters.

- accessJwt now stored as a secure connection parameter.

- Updated setheader policy: Authorization: Bearer {{accessJwt}}

troystaylor commented:
Hi Dan,
Please continue to test the actions using a connection made with the current connection parameters - I believe you will find that using the accessJwt as a connection parameter would lead to the action only being successful for the 60 seconds after the connection is made. My previous statement that {{accessJwt}} be set as a connection parameter was more about the syntax you were using, not that it was a suggestion for authentication to be implemented that way. Again, I have previously provided to you custom code to handle authentication (that works) and I don't think your method will ever be successful given the current limitations with what Microsoft allows for authentication methods - which I'm happy to be wrong about if proven otherwise.

krautrocker commented:
I see, I misunderstood you there. Let me take another look. Thanks, @troystaylor

krautrocker committed:
- Updated apiProperties.json to use query string parameters for authentication instead of headers.

- Added C# script to integrate with apiProperties authentication.

- paconn validated, no errors.

krautrocker commented:
Alrighty, took a sec but let's see how this tests out. Stay tuned and thank you!

CalvinTPark commented on independent-publisher-connectors/Bluesky Social/script.csx:
```diff
@@ -0,0 +1,170 @@
+                _logger.LogError("404 Not Found: The requested resource does not exist.");
+                return BuildErrorResponse(HttpStatusCode.NotFound, "The requested resource was not found.");
+
+            case (HttpStatusCode)529:
```
@krautrocker, maybe this should be HTTP status code `429`, to align with [Rate Limits
 | Bluesky](https://docs.bsky.app/docs/advanced-guides/rate-limits#:~:text=http%20429%20(%22too%20many%20requests%22)%20response%20status%20code) / [429 Too Many Requests - HTTP | MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429)?

krautrocker committed:
- Changed 529 error code to match 429, check docs: https://docs.bsky.app/docs/advanced-guides/rate-limits

certificationteam commented:
Hello @krautrocker, 

Is this ready for review and certification?

krautrocker commented:
> Hello @krautrocker, 
> 
> 
> 
> Is this ready for review and certification?

Yes - ready for review and ready to make any corrections, where necessary.

krautrocker committed:
- Fixed issue with required constraints data type in apiProperties - should be "true" and not true

krautrocker commented:
@vmanoharas I've been trying to push this fix through but having some issues with the OpenVet process. Has anything come through on your end?

vmanoharas reviewed (dismissed):
Dear Partner

I hope you are doing well.
Congratulations, your pull request is approved. We will move forward with the certification process. We are preparing your connector for production deployment and will queue up for the next deployment schedule. Once your connector onboards the next deployment schedule, it starts to deploy your connector in our production environments that typically takes 3-4 weeks.
Please let us know if you have any questions.
Thank you very much for working with us.

vmanoharas commented:
certify-connector

vmanoharas commented:
certify-connector

vmanoharas commented:
Hi @krautrocker, 

Please follow the below documentation and upload connector package zip file.

https://learn.microsoft.com/en-us/connectors/custom-connectors/certification-submission-ip#step-6-go-through-connector-validation-and-certification-process

Please refer this PR for example.  In this PR @troystaylor uploaded zip file, please upload only zip file not other files. 

https://github.com/microsoft/PowerPlatformConnectors/pull/3908/files

troystaylor commented:
> Hi @krautrocker,
> 
> Please follow the below documentation and upload connector package zip file.
> 
> https://learn.microsoft.com/en-us/connectors/custom-connectors/certification-submission-ip#step-6-go-through-connector-validation-and-certification-process
> 
> Please refer this PR for example. In this PR @troystaylor uploaded zip file, please upload only zip file not other files.
> 
> https://github.com/microsoft/PowerPlatformConnectors/pull/3908/files

@vmanoharas Please confirm the previously agreed plan that the zip file is in addition to the connector artifacts.

krautrocker committed:
- Parameter names are inconsistent between apiProperties, apiDefinition and script.csx
- Security settings need to be removed from the definitions
- Security settings need to be removed from all actions

- Added a policy to properties for a password parameter

- Updated definitions to include all parameters - pdshost, identifier and password

- Updated connectionProperties to include connection settings block

- Custom script parameters blueskyid and blueskypassword needed to be changed to identifier and password, respectively. The other names returned an error when running straight in Swagger.

krautrocker commented:
Will do whatever is required - I do need to get acquainted with the new certification process.

I've submitted some additional changes:

- Parameter names are now consistent between apiProperties, apiDefinition and script.csx
- Security settings needed to be removed from the definitions
- Security settings needed to be removed from all actions
- Added a policy to properties for a password
- Updated definitions to include all parameters - `pdshost`, `identifier` and `password`
- Updated connectionProperties to include connection settings block
- Changed custom script parameters `blueskyid` and `blueskypassword` to `identifier` and `password`, respectively. The other names returned an error when running straight in Swagger, I believe those names were causing a binding issue

Some necessary changes. @troystaylor I'm actually going to switch over to using your C# script. I had a hard time testing this with my code - it's been a number of months since I last looked at this - so I switched over to yours to try to get this working. I keep running into an error around certain properties "not available in the connection parameters" when I test the `CreateSession` action in the Power Automate test UI.

```
{
  "error": {
    "code": 500,
    "message": "Property with name identifier is not present in connection parameters.",
    "source": "ba85694e-e98f-e53f-80ac-26fcac62f51d.19.custom.usa002.azure-apihub.net",
    "path": "",
    "policyId": "",
    "clientRequestId": "bbd6e51a-7903-4626-b4d6-192aa45f67f3",
    "dreeSet": "true"
  }
}
```

A test in Swagger mode, however, is successful.

When I use the action in a flow, I get a 529 after multiple retries.

Something I'm overlooking, I hope.

🤷

krautrocker committed:
- Modified visibility for authFactorToken in CreateSession action

- paconn validated, no errors

krautrocker committed:
- Attribution to troystaylor for C# script, replaced own version

krautrocker committed:
- Updated operationId in custom code

krautrocker committed:
- Added privacy policy link

vmanoharas commented:
> > Hi @krautrocker,
> > Please follow the below documentation and upload connector package zip file.
> > https://learn.microsoft.com/en-us/connectors/custom-connectors/certification-submission-ip#step-6-go-through-connector-validation-and-certification-process
> > Please refer this PR for example. In this PR @troystaylor uploaded zip file, please upload only zip file not other files.
> > https://github.com/microsoft/PowerPlatformConnectors/pull/3908/files
> 
> @vmanoharas Please confirm the previously agreed plan that the zip file is in addition to the connector artifacts.
@troystaylor, no need to add connector artifacts, just a connector package zip file is enough.

vmanoharas commented:
> > Hi @krautrocker,
> > Please follow the below documentation and upload connector package zip file.
> > https://learn.microsoft.com/en-us/connectors/custom-connectors/certification-submission-ip#step-6-go-through-connector-validation-and-certification-process
> > Please refer this PR for example. In this PR @troystaylor uploaded zip file, please upload only zip file not other files.
> > https://github.com/microsoft/PowerPlatformConnectors/pull/3908/files
> 
> @vmanoharas Please confirm the previously agreed plan that the zip file is in addition to the connector artifacts.

@troystaylor, just connector package zip file is enough, no need to upload connector artifacts.

troystaylor commented:
@vmanoharas I understand it is enough, but the point of having this repo is to be able to view the contents of the artifacts and track line changes. Please have whomever is your new manager send me an email.

krautrocker committed:
- Adding ConnectorPackage.zip file for certification.

krautrocker commented:
@vmanoharas

1.) Structure validated via ConnectorPackageValidator:

<img width="1339" height="56" alt="image" src="https://github.com/user-attachments/assets/c1c1e74a-03e1-434c-b0aa-95d6780e7551" />

2.) Uploaded package as part of pull request, zip file only.

vmanoharas commented:
Hi @troystaylor, @krautrocker,  Yes, this package is sufficient. According to the new process, validation should run automatically. However, due to permission issues with new team members, the pipeline is currently not executing. Please allow a couple of days to resolve this issue. We do have approval to expedite the deployment, and once the pipeline is fixed, we will complete the deployment within three days.

vmanoharas commented:
certify-connector

vmanoharas commented:
certify-connector

vmanoharas commented:
certify-connector

certificationteam commented:
Congratulations — your certification request has been approved. Your connector will be included in the upcoming deployment schedule for production release and the supporting documentation will be published. This end-to-end process typically takes 3–4 weeks.

Many Thanks,
Copilot Connector Certification Team

vmanoharas reviewed (approved):
Hello Partner,
I hope you are doing well.
Congratulations, the deployment of your connector and documentation page was already completed, 
Thank you very much for working with us.
